src/lattices/abstract.jl

```diff
@@ -0,0 +1,66 @@
+"""
+Abstract definition of a lattice.
+"""
+abstract type AbstractLattice end
+
+# AbstractLattice interface: mandatory
+# TODO: This needs to be updated. There must be a general way to access sites and bonds.
+"""
+    length(l::AbstractLattice)
+
+Number of lattice sites.
+"""
+@inline Base.length(l::AbstractLattice) = error("Lattice $(typeof(l)) doesn't implement `length`.")
+
+
+
+
+
+
+# AbstractLattice interface: optional
+"""
+    neighbors(l::AbstractLattice, drop_repeats::Val{true})
+    neighbors(l::AbstractLattice, drop_repeats::Val{false})
+
+Returns an iterator over bonds, given as tuples (source index, target index). If
+`drop_repeats = Val(false)` bonds are assumed to be directed, i.e. both
+`(1, 2)` and `(2, 1)` are included. If `drop_repeats = Val(true)` bonds are
+assumed to be undirected, i.e. `(1, 2)` and `(2, 1)` are assumed to be
+equivalent and only one of them will be counted.
+"""
+@inline function neighbors(l::AbstractLattice, drop_repeats::Val{false})
+    (
+        [source, target] for source in 1:length(l)
+            for target in l.neighs[:, source]
+    )
+end
+@inline function neighbors(l::AbstractLattice, drop_repeats::Val{true})
+    (
+        l.bonds[i, 1:2] for i in 1:size(l.bonds, 1)
+    )
+end
+"""
+    neighbors(l::AbstractLattice, site_index::Integer)
+
+Returns a list of site indices neighboring the specified `site_index`.
+"""
+@inline function neighbors(l::AbstractLattice, site_index::Integer)
+    l.neighs[:, site_index]
+end
+
+
+# Typically, you also want to implement
+
+#     - `neighbors_lookup_table(lattice)`: return a neighbors matrix where
```

Comment:
More important than getting the neighbors as iterators is getting them as a lookup table (since we want to minimize neighbor lookup costs during Monte Carlo).

Reply:
I did some benchmarks. (with a dummy lattice, should be equivalent)

* `for trg in lattice.neighs[:, src]` is just as fast as `for trg in neighbors(lattice, src)`
* `for (src, trg) in neighbors(lattice, Val(false))` is 4x slower than `for src in 1:N; for trg in neighbors(lattice, src)`, but I think we only use this method to generate the checkerboard.
* `neighs[1] + neighs[2] + ...` with `neighs = neighbors(lattice, src)` is about 2.5x slower than indexing `lattice.neighs` directly, but
* `neighbors(lattice, src, 1) + neighbors(lattice, src, 2) + ...` is again just as fast, with `@inline neighbors(lattice, src, i) = lattice.neighs[i, src]`

For the Ising model all the general methods use either `neighbors(lattice, site_index)` (local, global updates) or `neighbors(lattice, Val(true)` (energy). The former performs fine, the latter doesn't use a lookup table.
Only the methods specialized on the `SquareLattice` suffer some performance penalties here, since there is no `neighbors(lattice, src, i)`. But I would argue that it's fine for those methods to just access `lattice.neighs` directly, because they only work for one specific lattice we implement.

As you mentioned below, `hopping_matrix` uses `neighbors(lattice, Val(false))`. I'm not too bothered by this, as it's only called once. But we can easily change this to use `neighbors(lattice, site_index)` to get better performance.

Reply:
```julia
using BenchmarkTools, MonteCarlo
import MonteCarlo: neighbors, HoneycombLattice, AbstractLattice

l = HoneycombLattice(20)

eachsite(l::AbstractLattice) = Base.OneTo(length(l))

@inline neighbors(lattice, src, i) = lattice.neighs[i, src]

function f_neighbors_site(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for trg in neighbors(l,src) # allocates
            x += trg
        end
    end
    x
end

function f_neighbors_site_i(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for i in 1:size(l.neighs, 1)
            trg = neighbors(l, src, i)
            x += trg
        end
    end
    x
end

function f_table_lookup(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for i in 1:size(l.neighs, 1)
            trg = l.neighs[i,src]
            x += trg
        end
    end
    x
end

######## Microbenchmarks
@btime neighbors(l,1);
@btime $(l.neighs)[:,1];

# julia> @btime neighbors(l,1);
#   57.580 ns (1 allocation: 112 bytes)

# julia> @btime $(l.neighs)[:,1];
#   40.482 ns (1 allocation: 112 bytes)

@btime neighbors(l,1)[3];
@btime $(l.neighs)[3,1];

# julia> @btime neighbors(l,1)[3];
#   80.249 ns (2 allocations: 128 bytes)

# julia> @btime $(l.neighs)[3,1];
#   1.799 ns (0 allocations: 0 bytes)


######## Context benchmarks
@btime f_neighbors_site($l);
@btime f_table_lookup($l);
@btime f_neighbors_site_i($l); # as fast as f_table_lookup because it is literally identical after inlining

# julia> @btime f_neighbors_site($l);
#   63.899 μs (1600 allocations: 175.00 KiB)

# julia> @btime f_table_lookup($l);
#   6.860 μs (0 allocations: 0 bytes)

# julia> @btime f_neighbors_site_i($l);
  # 6.880 μs (0 allocations: 0 bytes)
```

Reply:
Currently, the field `neighs` of all lattices is a lookup table. Accessing it is as fast as it gets. Generally, we don't want anyone to access fields directly. First of all, the lattice implementation might change at some point but more importantly we want to support lattices from other libraries which might not have an internal nearest neighbors table. Therefore, we define an `AbstractLattice` interface. The question is what mandatory and optional methods do we want as the interface specification.

Proposal:

**Mandatory**

Every lattice should implement those, otherwise no one could use it on the model side for anything.
* `length`: number of lattice sites
* `ndims`: number of dimensions
* `neighbors(lattice[, directed=Val(false)`: Iterator over `(src,trg)` tuples (or `src=>trg` pairs?). If `directed=Val(true)` both `(src,trg)` and `(trg,src)` will be generated.
(This function might be called `bonds(lattice)` in other lattice libraries)

**Optional**

* `neighbors(lattice, site)`: Iterator over neighbors of a given site.
* `neighbor(lattice, site, i)`: Return the site index of the `i`th neighbor of a site
* `neighbors_lookup_table(lattice)`: generate a table for O(1) nearest neighbor lookup
* perhaps `nneighbors(lattice, site)`: number of neighbors of a given site
* perhaps `name`: name of the lattice

We could (and should) provide generic fallback implementation of the optional methods, (based on the mandatory `neighbors(lattice)` iterator).

Reply:
Currently, the field `neighs` of all our lattices is a lookup table. Accessing it is as fast as it gets (personally I'd always use a lookup table for Monte Carlo because I cannot be sure that the accessor methods of the lattice are efficient). Generally, we don't want anyone to access fields directly. First of all, the lattice implementation might change at some point but more importantly we want to support lattices from other libraries which might not have an internal nearest neighbors table.

Having a generic `neighbors_lookup_table(lattice)`, even if inefficient, would be a great way to support almost any lattice in `O(1)` speed (at least for nearest neighbors) independent of its implementation.

Reply:
My dummy lattice is

```julia
struct Lattice
    N::Int64
    neighs::Matrix{Int64}
end

@inline function neighbors(l::Lattice, site_index::Integer)
    l.neighs[:, site_index]
end

N = 10
neighs = rand(Int64, 4, N)
l = Lattice(N, neighs)
```

With it I get:

```julia
julia> @btime neighbors(l,1);
  26.339 ns (1 allocation: 112 bytes)

julia> @btime $(l.neighs)[:,1];
  25.269 ns (1 allocation: 112 bytes)
```

I get `34ns` and `26ns` respectively with your benchmarks though. Defining `neighbors` for the concrete lattice gets it down to `26ns` as well.

Reply:
```julia
function f_table_lookup(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for i in 1:size(l.neighs, 1)
            trg = l.neighs[i,src]
            x += trg
        end
    end
    x
end
```

My version uses `for trg in l.neighs[:, src]`, which causes allocations and slows it down to be on par with `for trg in neighbors(l, src)`. With a concrete `HoneycombLattice` in `neighbors()` I'm getting `39.67 μs`, `5.03 μs`, `4.92 μs` for the last three tests, so still a significant slowdown. (Same number of allocations)

Reply:
Your `neighbors(l,1)` and `l.neighs[:,1]` allocate. This isn't necessary. We shouldn't have to allocate anywhere in a model to access/iterate over neighbors.

We should always use `l.neighs[i,1]` or better `neighbor(l, 1, i)` (since accessing fields directly is bad). (We should change this in IsingModel.jl.

Reply:
> * `neighbors_lookup_table(lattice)`: generate a table for O(1) nearest neighbor lookup

I assume you want to do something like `model.neighs = neighbors_lookup_table(lattice)`. I'd rather keep the lattice and the model more seperate and interface through `neighbor(lattice, site, i)`. 
That would require

> * perhaps `nneighbors(lattice, site)`: number of neighbors of a given site

but we should find a different name, one not so easily confused with `neighbors()`.

> * perhaps `name`: name of the lattice

I'm not sure why - the name of the lattice is the type of the lattice already.

Reply:
> I assume you want to do something like `model.neighs = neighbors_lookup_table(lattice)`. I'd rather keep the lattice and the model more seperate and interface through `neighbor(lattice, site, i)`.

I want both! If I implement a model I can choose whether I want to use a lookup table (which is guaranteed to be fast) or whether I want to rely on accessor functions of the lattice, like `neighbor(lattice, site, i)`. My personal preference is the former but I'd leave it up to the taste of the person implementing the model.

The reason why I do not want to only have the latter option is that we want to support lattices from external sources like Lattices.jl or LatticePhysics.jl and we might not have full control over the efficienty of accessor methods. The lookup table on the other hand, even if it is costly to generate at first, is always fast.

What we choose for the built-in models, doesn't matter that much. Perhaps we should demonstrate both approaches?

Reply:
> We should always use `model.neighs[i,1]` or `neighbor(l, 1, i)` (We should change this in IsingModel.jl). The former, having the lookup table on the model side, has the advantage that access is guaranteed to be fast indepentend of the lattice's implementation of `neighbor(l,1,i)`.

We should just pick one direction and go with it. We can change it later if needed.

Reply:
Btw, what would you need `ndims` for?

Reply:
I guess we don't need it. Same for `name`.

Reply:
BTW, I dropped the `AbstractCubicLattice` locally. This should be a trait rather than an abstract type. Also, since we don't want to implement `neighbors(lattice, directed)` for all built-in lattices independently, I'm implementing traits for that as well. Something like this:

```julia
# Abstract trait
abstract type LatticeProperties end

"""
    HasNeighborsTable trait

Indicates that a lattice `l` has a neighbors lookup table as a field `l.neighs`.

A neighbors lookup table is a `Matrix{Int}` of size `(nneighbors, nsites)` such that
`l.neighs[i,j]` gives the site index of the `i`th neighbor of site `j`.

A lattice should implement `has_neighbors_table(::MyLattice) = HasNeighborsTable()` to indicate the trait.
"""
struct HasNeighborsTable <: LatticeProperties end

"""
    HasBondsTable trait

Indicates that a lattice `l` has a bonds table as a field `l.bonds`.

A bonds table is a `Matrix{Int}` of size `(nbonds, 3)` where the columns mean
`src`, `trg`, `type`, respectively.

A lattice should implement `has_bonds_table(::MyLattice) = HasBondsTable()` to indicate the trait.
"""
struct HasBondsTable <: LatticeProperties end

has_neighbors_table(::AbstractLattice) = nothing
has_bonds_table(::AbstractLattice) = nothing

@inline neighbors(l::AbstractLattice, directed::Val{true}) = _neighbors(has_neighbors_table(l), l, directed)
@inline neighbors(l::AbstractLattice, directed::Val{false}) = _neighbors(has_bonds_table(l), l, directed)

@inline _neighbors(::Nothing, l::LT, directed::V) where {LT<:AbstractLattice, V<:Val} =
    error("Lattice $(LT) doesn't implement `neighbors(::$(LT), ::$V`")

@inline function _neighbors(::HasNeighborsTable, l::AbstractLattice, directed::Val{true})
    (
        (src, trg) for src in 1:length(l)
            for trg in l.neighs[:, src]
    )
end
@inline function _neighbors(::HasBondsTable, l::AbstractLattice, directed::Val{false})
    (
        l.bonds[i, 1:2] for i in 1:size(l.bonds, 1)
    )
end
```

Reply:
I wonder if this is more work than it's worth. For classical mc people probably implement their own models and lattices and tweak them as needed, rather than following whatever (lattice) interface we define. And for dqmc the lattice only matters for the setup, right?

Reply:
> I wonder if this is more work than it's worth.

Maybe.

> For classical mc people probably implement their own models and lattices and tweak them as needed, rather than following whatever (lattice) interface we define.

The idea is that people only have to implement their model but could use/build upon a pre-defined lattice (from Lattices.jl, LatticePhysics.jl, or built-in). They don't really have to care about the `AbstractLattice` interface other than to see how they get the information from a lattice they need. In that sense, our `AbstractLattice` is basically just a synchronization of different `AbstractLattice` definitions and concrete lattice implementations out there. Sure, we could also just say, "if you use lattice library X look at their documentation for conventions and accessor functions". However, I think it's good to have a common interface. This is also why I like functions like `neighbors_lattice_table(::AbstractLattice)`. After I call this function the lattice, independent of its origin, is just a lookup table that I know how to use.

> And for dqmc the lattice only matters for the setup, right?

Not quite, one also has to calculate the energy change in the bosonic sector for Metropolis. However, I agree that this is hardly a bottleneck in DQMC.

Reply:
Of course, if Lattices.jl or LatticePhysics.jl were already mature/established, we wouldn't have to any of this. We could just support that single library and copy their conventions.